src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -267,70 +274,98 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
 {
     is_execution_initialized = true;
 
+    /// Allocate CPU slots from concurrency control
+    constexpr size_t min_threads = 1;
+    slots = ConcurrencyControl::instance().allocate(min_threads, num_threads);
+    size_t use_threads = slots->grantedCount();
+
     Queue queue;
     graph->initializeExecution(queue);
 
-    tasks.init(num_threads, profile_processors, trace_processors, read_progress_callback.get());
+    tasks.init(num_threads, use_threads, profile_processors, trace_processors, read_progress_callback.get());
     tasks.fill(queue);
+
+    std::unique_lock lock{threads_mutex};
+    threads.reserve(num_threads);
+}
+
+void PipelineExecutor::spawnThreads()
+{
+    while (auto slot = slots->tryAcquire())
+    {
+        std::unique_lock lock{threads_mutex};
+        size_t thread_num = threads.size();
+
+        /// Count of threads in use should be updated for proper finish() condition.
+        /// NOTE: this will not decrease `use_threads` below initially granted count
+        tasks.upscale(thread_num + 1);
```

Comment:
I think we can just make a method `tasks.upscale()` without arguments which just increases `use_threads` under mutex. Why do we need some extra argument? :)

Reply:
It is due to the fact the init()/fill() must use use_threads that is equal to initially granted amount of threads to distribute tasks properly, but spawnThreads() is used even for those initially granted threads